cache/blobs.go

```diff
@@ -455,24 +485,34 @@ func ensureCompression(ctx context.Context, ref *immutableRef, comp compression.
 
 		// First, lookup local content store
 		if _, err := ref.getBlobWithCompression(ctx, comp.Type); err == nil {
-			return nil, nil // found the compression variant. no need to convert.
+			return nil, nil // found the compression variant. no need to ensure compression.
 		}
 
-		// Convert layer compression type
-		if err := (lazyRefProvider{
-			ref:     ref,
-			desc:    desc,
-			dh:      ref.descHandlers[desc.Digest],
-			session: s,
-		}).Unlazy(ctx); err != nil {
-			return nil, err
-		}
-		newDesc, err := layerConvertFunc(ctx, ref.cm.ContentStore, desc)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to convert")
+		var newDesc *ocispecs.Descriptor
+		if errors.Is(cvtErr, errdefs.ErrNotImplemented) {
+			// For the source compression type which haven't provide decompression (for converting)
+			// ability, re-compute diff to ensure ref has the specified target compression type.
+			newDesc, err = computeBlob(ctx, ref, comp, s)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to compute blob for %q", comp.Type)
+			}
```

Comment:
If the main blob (returned from `ref.ociDesc()`) is nydus, does this trigger the whole diff computation every time a compression variant is requested? Maybe we should deduplicate this .